infra/DockerRepoManager.py

```diff
@@ -0,0 +1,189 @@
+
+    drm = DockerRepoManager('curl')
+    current_commit = drm.get_image_commit()
+    drm.set_image_commit('df26f5f9c36e19cd503c0e462e9f72ad37b84c82')
+
+"""
+from helper import _build_image
+from helper import _check_project_exists
+from helper import _get_dockerfile_path
+from helper import _is_base_image
+from RepoManager import RepoManager
+import os
+
+
+class NoRepoFoundException(Exception):
+  """Occurs when the bisector cant infer the main repo."""
+  pass
+
+
+class DockerRepoManagerException(Exception):
+  """When there is a docker error of execution"""
+  pass
+
+
+class DockerRepoManager(RepoManager):
+  """Class to manage a git repo inside of a docker continer.
+
+  Attributes:
+    docker_image: The name of the docker image that is being modified
+    project_name: The name of the project associated with the docker image
+  """
+
+  docker_image = ''
+  project_name = ''
+  src_on_image = ''
+  TEMP_IMAGE_NAME = 'temp_container'
+
+  def __init__(self, project_name):
+    """Inits the DockerRepoManager class.
+
+      Args:
+        project_name: The name of the project that is associated with the image
+    """
+    self.project_name = project_name
+    self.docker_image = self._get_docker_image_name()
+    repo_url = self._infer_main_repo()
+    self.src_on_image = os.path.join('/src', project_name)
+    super().__init__(repo_url)
+    if self.repo_name != project_name:
+      raise DockerRepoManagerException(
+          'Error, the project name must be the same as the ' +
+          'git repo name but are %s and %s' % (project_name, super().repo_name))
+
+  def set_image_commit(self, commit):
+    """Creates a docker image with a specified commit as its source.
+
+    Args:
+      commit: The SHA the source is to be checked out at
+
+    Raises:
+      DockerRepoManagerException: when the commit is not successfully
+      mounted to the image
+    """
+
+    #Remove old temp container
+    stop_command = ['docker', 'stop', self.TEMP_IMAGE_NAME]
+    self._run_command(stop_command)
+    remove_command = ['docker', 'container', 'rm', self.TEMP_IMAGE_NAME]
+    self._run_command(remove_command)
+
+    # Remove all previous images
+    rmi_command = ['docker', 'rmi', self.docker_image]
+    self._run_command(rmi_command)
+
+    # Build builder image
+    _build_image(self.project_name)
+    self.checkout_commit(commit)
+    mount_command = [
+        'docker', 'create', '--name', self.TEMP_IMAGE_NAME, self.docker_image
+    ]
+    self._run_command(mount_command)
+
+    # Start the container to be modified
+    start_container = [
+      'docker', 'start', self.TEMP_IMAGE_NAME
+    ]
+    self._run_command(start_container)
+
+    # Remove outdated source repo from container
+    remove_command = [
+        'docker', 'exec', self.TEMP_IMAGE_NAME, 'rm', '-rf', self.src_on_image
+    ]
+    self._run_command(remove_command)
+ 
+    # Copy updated source repo to container
+    copy_command = [
+        'docker', 'cp',
```

Comment:
why do we need to create containers with this copied? how about just using -v to mount the sources from the host rather than this complicated steps of starting a container and then copying the sources?

Reply:
I think copying is better than mounting.
Mounting the source directory will allow changes made during builds to persist to the next build. I think I've encountered projects that won't rebuild unless `make clean` is run. Copying avoids any issue with this.

Reply:
Sure, but I believe most (all?) source control systems allow you to clean your git dir with a single command:

e.g. `git clean -fxd` which should also include ignored files to bring your git repo back to a pristine state. I think this is much cleaner than all of these steps to manage temporary containers.

Reply:
I don't love the temporary images either. I guess if it works as well I support mounting over copying.